### ISSUE TITLE
IntelliJ / Webstorm / IDEA sourcemap debugging compatibility?

### ISSUE THREAD

**Issue by anyong** [substack/node-browserify#1233]
I use Intellij IDEA and am unable to get debugging working with browserify's sourcemaps. The people at JetBrains think it's a browserify problem, as there have been a [couple](https://devnet.jetbrains.com/message/5540299#5540299) of [issues](https://youtrack.jetbrains.com/issue/WEB-14454) posted about that already. I have the same issue as the second link - maps are found and correctly linked, but breakpoints not hit. Does anyone here think this is a browserify issue or a JetBrains issue? Probably the right person on each side would be able to figure this one out.

**Comment by terinjokes** [comment]
How do you setup IntelliJ in a Browserify project? I've been known to use it on occasion, but have never been successful in getting Chrome and IntelliJ to talk to each other.

I'm also happy to chat with anyone from JetBrains about this issue if they see this issue.

**Comment by develar** [comment]
Hi, I am from JetBrains. I am working now to support Browserify again (I supported Browserify sourcemaps in 2013 year).

We have fixed number of generic (unrelated to Browserify) issues on our side (fixed in WebStorm 11 / 142 branch).

But there are two Browserify issues:
1) minor: incorrect data uri. Why Browserify uses "charset:utf-8:" instead of "charset=utf-8" (as it should be according to spec)? In any case Google Chrome supports such format, so, we have supported it.

2) major: Browserify generates incorrect source paths.

Please see sample project: https://github.com/develar/browserify-sample

Sources must be relative to "out" dir, not to root ("/") dir. But Browserify generates "src/app.js". So, even if you configure mappings (map \<project root> to http://localhost:8090/browserify and \<project root>/out to http://localhost:8090/browserify/out), WebStorm cannot resolve sources. And, so, breakpoints don't work.

Workaround: map "src" folder to http://localhost:8090/browserify/out/src. (here localhost:8090 is a webstorm built-in server, 8090 custom port).

From my experience, any your attempt to generate source path correctly using standard "file" or "http" protocol will be failed. Any approach to compute correct source path has drawbacks. 

* If you don't specify protocol (file or http), it means, according to sourcemap spec (https://docs.google.com/document/d/1U1RGAehQwRypUTovF1KRlpiOFze0b-_2gc6fAH0KY0k/edit#heading=h.75yo6yoyk7x5), "relative to the SourceMap (like resolving script src in a html document)." HTTP scheme will be used.

 As you can imagine, it leads to magic unresolvable URLs (http://localhost:8090/browserify/out/src).
 Unresolvable because in most cases sources will not be available over HTTP.
 Magic because IDE will transform HTTP URL to local path somehow (magic). 

* If you specify absolute path without "file://" (as you did in 2013, as far I remember) — it is ambiguous — path relative to local file system root, or to http host?
* If you specify absolute path with "file://" it is not suitable 
  * for build tools. For example,  gulp sourcemaps / uglify cannot merge source path correctly with another (for example, user can specify some custom source root and it will be transformed). 
  * absolute path is not suitable because you cannot debug app on another machine (path must be relative to project (not IDEA project, but may be gulp/webpack/build tool project) root).
* if you specify relative path to script file (as you do now (incorrectly, but anyway)) — it is unreliable approach because no direct mapping between compiled script file (http, different project configuration) and source file (local FS).

So, I strongly recommend you to use webpack approach. webpack solved this problem elegantly — webpack uses own scheme "webpack://". For example, any file inside webpack roots prepended with "webpack:///./". So, there is the only straightforward, unambiguous way to resolve sources. Debugging in the Google Chrome still working, because source contents provided in any case.
Well, yes, webpack approach was a reason why we supported browserify much earlier than webpack (2013 vs 2015 :)). But now (since WS 11, see https://docs.google.com/document/d/19rmW6PGpIIx57ntk4M_1x6F0t-0qO2weCtNKWQJaemA/edit) any complex webpack application could be debugged, but browserify not.

**Comment by jmm** [comment]
@develar Thanks for this detailed analysis.

By...

> root ("/") dir

...you mean the base dir for the bundle, right?

> 2) major: Browserify generates incorrect source paths.

I'm not sure it's accurate to say that the paths are incorrect. Probably the most sensible thing in general is for them to be relative to the bundle base dir. It wouldn't generally make sense for them to be relative to the output directory -- that's arbitrary. Browserify really has no concept of an output directory except as a convenience for the CLI to write the file to somewhere (and even that is only sometimes with the CLI), and the output directory is changed merely by moving the bundle file.

I don't want browserify automatically putting `file://` or `webpack://` or any such thing in the source maps. It sounds to me like maybe the best solution here is to just expose `sourceRoot` as an option, which was just requested in #1304. With `sourceRoot` you could instruct your users what value to set (e.g. `browserify://<whatever>`) and that would allow you to construct the paths you need, correct?


> 1) minor: incorrect data uri. Why Browserify uses "charset:utf-8:" instead of "charset=utf-8" (as it should be according to spec)?

I believe that comes from `combine-source-map`.

**Comment by develar** [comment]
> you mean the base dir for the bundle, right?

I mean current working directory (where I run "browserify --debug src/app.js -o out/bundle.js"). Project base dir. As far I understand, in Browserify terms, yes, "base dir for the bundle".

> I'm not sure it's accurate to say that the paths are incorrect. Probably the most sensible thing in general is for them to be relative to the bundle base dir.

The problem — if you don't use custom scheme (webpack:// for example) and use standard path (http currently) — you should follow spec. Spec defines it "relative to the SourceMap (like resolving script src in a html document)" (https://docs.google.com/document/d/1U1RGAehQwRypUTovF1KRlpiOFze0b-_2gc6fAH0KY0k/edit#heading=h.75yo6yoyk7x5).

You violate the spec without any indication. I am totally agree with you — "the most sensible thing in general is for them to be relative to the bundle base dir".

But without some indication no one knows about it. Without indication standard resolving (and standard canonicalisation and so on) will be applied and, obviously, will be failed. Spec just silently violated (browserify --debug src/app.js -o out/bundle.js — so, path to source should be relative to out dir).

In case of webpack it is very easy to map "webpack://./" (here "." means "bundle base dir"). It is clear, that URL webpack:// is a custom and must not be processed according to spec (canonicalization, resolving).

> With sourceRoot you could instruct your users what value to set (e.g. browserify://<whatever>) and that would allow you to construct the paths you need, correct?

Yes. But, I think, browserify should produce debuggable sourcemaps by default, without any custom configuration (IDE configuration is another story).

> I believe that comes from combine-source-map.

May be. My sample project (https://github.com/develar/browserify-sample) is very easy and I use only browserify CLI to compile.

**Comment by BigstickCarpet** [comment]
Not sure if this helps or not, but I wrote a Browserify plugin to "fix" the sourcemap paths.  It's been working well for my team on several projects.   It still doesn't work with WebStorm's debugger, but at least the paths are correct in the Chrome Dev Tools, so debugging there is easy.

https://www.npmjs.com/package/sourcemapify

**Comment by develar** [comment]
@BigstickCarpet I tested it — doesn't work with current WebStorm's debugger due to "data uri" issue. But work with WebStorm 11 (browserify src/app.js -o out/bundle.js --debug -p [ sourcemapify --root "../" ]).

**Comment by jmm** [comment]
@develar 
> I mean current working directory (where I run "browserify --debug src/app.js -o out/bundle.js"). Project base dir. As far I understand, in Browserify terms, yes, "base dir for the bundle".

Yes, browserify's base dir will either be the value of the `basedir` option or default to `process.cwd`.

> The problem — if you don't use custom scheme (webpack:// for example) and use standard path (http currently) — you should follow spec. Spec defines it "relative to the SourceMap (like resolving script src in a html document)".

I understand where you're coming from, but what you're talking about is outside the scope of the source map spec, which is written in the context of HTTP servers and user agents. The spec merely defines how a user agent can construct a URL to request a source file from. Furthermore, browserify generates source maps with `sourcesContent`, which for the use case that source maps were designed for obviates the need to even locate the source files.

> You violate the spec without any indication.

I don't believe that's accurate. This problem, as well as others I've seen, stems from wanting to use source maps for something they weren't designed for (e.g. referring to local file system paths). I'm not saying source mapping can't or shouldn't accomodate the use case you're talking about, possibly even as currently spec'ed, but it's not the use case the spec describes and that has important implications.

The directory a browserify bundle is written to is totally arbitrary and there's no reason to presume that the value of the `-o` argument to a CLI command has any relationship to the values in the source map. That would enforce a rigid correspondence between local file system structure and later resolution behavior that not everyone wants. Many people don't want source maps leaking absolute local file system paths. And that would only apply to CLI commands with the `-o` argument anyway -- not to CLI commands that just do `browserify entry.js > somedir/bundle.js` or any API commands.

There may well be a case for making browserify's source map output more configurable though. Case in point, I believe `sourceRoot` would solve your problem, without forcing an abritrary connection to local file system structure on everyone.

> Yes. But, I think, browserify should produce debuggable sourcemaps by default, without any custom configuration (IDE configuration is another story).

The best you can do without configuration is choose sensible defaults, which is the current status. They're debuggable by default in the context the system was designed for -- browsers, that support `sourcesContent`.

> May be. My sample project (https://github.com/develar/browserify-sample) is very easy and I use only browserify CLI to compile.

Yes, I believe `combine-source-map` is the dependency browserify uses to generate its source maps, including the `charset` part.

BTW, please mark up code snippets, file names, etc. in your posts :) See [How to report issues with browserify](https://github.com/substack/node-browserify/wiki/How-to-report-issues-with-browserify) for more info.

**Comment by develar** [comment]
> as well as others I've seen, stems from wanting to use source maps for something they weren't designed for (e.g. referring to local file system paths). I'm not saying source mapping can't or shouldn't accomodate the use case you're talking about,

The spec defines how to resolve sources. It is clearly stated in the spec. Resolve not to local filesystem, but to absolute URL (if scheme is not defined, HTTP).

It the reason why JetBrains debugger and Google Chrome display generated browserify result in the same way: ![](https://dl.dropboxusercontent.com/u/43511007/s/Screen%20Shot%202015-06-17%20at%2020.59.32.png)

It doesn't matter that sourcemap spec defines unusable / unmaintainable / unwanted approach to specify and resolve sources. You must or generate absolute URL (I am totally agree — http or file protocols should be avoided — it is not maintainable and  "Many people don't want source maps leaking absolute local file system paths"), or generate relative path according to spec.

Because sourcemap is a standard and another tool, IDE in our case, is not aware about sourcemap producer. Typescript compiler, for example, follows the spec (path generated correctly, relative to out dir).

If you violates the spec, we cannot process sourcemap using standard way (according to spec). It leads to incorrect URLs. And as result to configure debugging you have to setup strange and unclear mappings (instead of map "bundle base dir" to source root URL, you have to explicitly map source root to incorrect/strange URL (e.g. `http://localhost:8090/browserify/out/src`)). ![screenshot of JS Debug RC](https://dl.dropboxusercontent.com/u/43511007/s/Screen%20Shot%202015-06-17%20at%2021.02.46.png)

Instead of just map "bundle base dir" to `webpack:///.` ([example](https://docs.google.com/document/d/19rmW6PGpIIx57ntk4M_1x6F0t-0qO2weCtNKWQJaemA/edit#heading=h.p0l0ykwhu2rr)).

So, my proposal:
Consider to use elegant webpack way by default to solve this problem: introduce custom URL scheme to prevent URL canonicalization and to indicate your tool (in case of webpack we don't have to support webpack specially — it just works (and without any user configuration on webpack side)). 

May be I am wrong, but currently usage of browserify requires knowledge about "sources relative to the bundle base dir" in any case. So, such scheme just simplify usage (any tool should be aware about it (devtools is out of scope because devtools doesn't work with real project sources)). 

Imagine — how we can detect that sourcemap produced by browserify and apply non-standard logic to resolve sources? I vote and ask you to fix it by default — it is not user-friendly to require specify "sourceRoot" explicitly (every user will encounter this issue and search the solution).

**Comment by jmm** [comment]
@develar As you read my response please keep in mind that I understand where you're coming from and I'd like to help you find a reasonable solution to your issue, but I disagree with some of your interpretations and proposed solutions.

> The spec defines how to resolve sources. It is clearly stated in the spec. Resolve not to local filesystem, but to absolute URL (if scheme is not defined, HTTP).

Yes, we're on the same page here: the values in question are spec'ed as URLs.

> It the reason why JetBrains debugger and Google Chrome display generated browserify result in the same way:

I'm not sure what you mean by that.

> It doesn't matter that sourcemap spec defines unusable / unmaintainable / unwanted approach to specify and resolve sources.

I'm not sure what you mean. You feel that the source map spec defines an "unusable / unmaintainable / unwanted approach" or that browserify violates the spec?

> You must or generate absolute URL (I am totally agree — http or file protocols should be avoided — it is not maintainable and "Many people don't want source maps leaking absolute local file system paths"), or generate relative path according to spec.

Again, I'm not really sure what you mean here. You don't have to embed absolute URLs in the source map. If you want a consumer to be able to locate the sources you have to coordinate the relevant values appropriately to make that possible.

> Typescript compiler, for example, follows the spec (path generated correctly, relative to out dir).

I can't stress this enough: that has nothing to do with following the spec. The spec doesn't say anything about "if your build tool writes output to dir `x`, paths in the source map must be relative to `x`". The spec only defines how the values are to be interpreted by the consumer. For your project or your tool you may want to write output to `x` and have the values in the source map be relative to `x`, but that's not universal.

> If you violates the spec, we cannot process sourcemap using standard way (according to spec). 

I disagree that there's violation of the spec. I believe that what you're talking about is outside the scope of the spec.

> Consider to use elegant webpack way by default to solve this problem: introduce custom URL scheme to prevent URL canonicalization and to indicate your tool (in case of webpack we don't have to support webpack specially — it just works (and without any user configuration on webpack side)).

I personally would oppose doing this without a compelling case for why it makes sense. What if someone wants a source map that works the way the spec intends -- so that the consumer (let's say a browser) can retrieve the sources via HTTP? Like you acknowledged:

> Because sourcemap is a standard and another tool, IDE in our case, is not aware about sourcemap producer.

Nor should it be: source mapping is supposed to be tool-agnostic.

> May be I am wrong, but currently usage of browserify requires knowledge about "sources relative to the bundle base dir" in any case.

That's always the case. The `sources` values are either absolute URLs or relative to some base URL. If you bundle files in `/somedir` and write the bundle to `/somedir/out`, even if you make the source map values relative to `out`, if you move the bundle to `/otherdir` the link is broken.

> Imagine — how we can detect that sourcemap produced by browserify and apply non-standard logic to resolve sources?

You don't need to detect that it's produced by `<insert tool here>` because there's a standard mechanism that accomplishes what you want: `sourceRoot`.

> I vote and ask you to fix it by default — it is not user-friendly to require specify "sourceRoot" explicitly (every user will encounter this issue and search the solution).

As I said, without configuration you can only choose a sensible default. It's not user friendly to default to one particular rigid correspondence between local file system paths and source map values that not everyone wants, or to populate the source map with a non standard URL scheme.

> (every user will encounter this issue and search the solution).

Every user of your tool. I'm sure you can see how that's not a good reason for changing the source map output to a niche format for everyone, right? And what about your users that use browserify without the CLI `-o` option? What is supposed to happen to them?

**Comment by develar** [comment]
Babel also generates correct sourcemaps, opposite to browserify (and browserify doesn't work again in my test project). Ok, it is up to you. We are waiting user feedback — https://youtrack.jetbrains.com/issue/WEB-14454#comment=27-1016776 

Thanks for clarification.

**Comment by jmm** [comment]
From your ["IntelliJ JavaScript debugging with source maps"](https://docs.google.com/document/d/19rmW6PGpIIx57ntk4M_1x6F0t-0qO2weCtNKWQJaemA/edit?pli=1#) document:
> (also, in the near future will be plugins for webpack and gulp to avoid manual configuration)

What does it mean by plugins? Does it actually mean plugins for webpack and gulp, or does it mean a plugin for your editor that's related to webpack / gulp?

**Comment by develar** [comment]
>> What does it mean by plugins? 
Yes, "actually mean plugins for webpack and gulp", not intellij platform plugins. We are going to write nodejs plugins to solve unrelated issues — providing mappings ("auto configuration") will not be main functionality.

**Comment by jmm** [comment]
If you don't want to instruct your users to set a `sourceRoot` option for browserify (if and when it becomes available, which I think will probably happen) then perhaps providing them with a browserify plugin would be an alternative.

**Comment by develar** [comment]
Yes, if issue will be actual (confirmed by user reports), some workaround, probably (may be, I will try to open another round to convince you using real projects :)), will be implemented. We are waiting user feedback.

**Comment by jmm** [comment]
@develar Ok. Nothing here is set in stone, but I can tell you there'd have to be a really compelling case made for defaulting to things like including any URI scheme in the output, or basing the source map output on an *output* directory.

**Comment by terinjokes** [comment]
@jmm The "compelling case" is ensuring users can debug their code using the tools they want.

**Comment by jmm** [comment]
@terinjokes Ok, so that's a compelling case for what specifically? There are 2 things to do:

1. Choose sensible, general purpose defaults.
1. Provide adequate configurability.

What specifically do you propose?

**Comment by raphw** [comment]
I (with some help) found a work-arround for successfully using browserify source maps from IntelliJ: http://stackoverflow.com/questions/31070112/enable-javascript-debugging-with-intellij-and-source-maps

Maybe this also helps to identify where to seek for a fix.

**Comment by nmccready** [comment]
First off thank you for browserify. I like it very much and there were reasons why we moved to this from webpack (mainly css sourcemap bugs).

I am not sure of this is the right place to be asking for feedback on this; but here it goes:

Is there any concept of something similar for sourcemaps like webpacks [devtool](http://webpack.github.io/docs/configuration.html#devtool) . I am mainly looking into this because currently our sourcemaps debug to straight coffee. While I love writing in coffee, debugging in it is a nightmare. 

In webpack the devtool: 'eval' options defaults the debugged sourcemaps to be in `js` which is what I desire. So my main goal is to find out what the pipeline if any to get my sourcemapped coffeescript to render as debuggable javascript in the browser.

**Comment by BigstickCarpet** [comment]
@nmccready - I ended up creating a [custom Browserify CLI tool](https://www.npmjs.com/package/simplifyify) for our team, and it has worked great for us.  It uses [Exorcist](https://www.npmjs.com/package/exorcist) to create external sourcemaps (`*.map` files), which work perfectly with WebStorm's debugger.

We're not using CoffeeScript, so I haven't tested our tool with that yet, but it _should_ work as long as you add CoffeeScript to the [`browserify.transform` field in `package.json`](https://github.com/substack/node-browserify#browserifytransform)

**Comment by nmccready** [comment]
@BigstickCarpet so would I be able to debug the coffee as js in seperate js files?

**Comment by BigstickCarpet** [comment]
You _should_ be able to.  We're able to debug separate js files in WebStorm without any problems.  But we're not using CoffeeScript, so I don't know if there are any "gotchas" involved in that

**Comment by nmccready** [comment]
10-4 I'll give it a shot ty very much.

**Comment by nmccready** [comment]
@BigstickCarpet what if I want use your api through gulp itself and not the cli is there any docs for that?
